src/server/index.js

```diff
@@ -109,96 +115,219 @@ const createStock = quote => {
   };
 };
 
-const calculateDateYearsInPast = years => {
-  const year = new Date().getFullYear() - years;
-  const date = new Date();
-  date.setFullYear(year);
-  return date;
+// Used for historical() data obtained using period 'd'
+const getDatesAndPrices = dailyData => {
+  let datesAndPrices = [];
+  dailyData.forEach(({
+    date,
+    close
+  }) => {
+    datesAndPrices.unshift({
+      date: formatDateForMaxStockData(date),
+      price: close
+    });
+  });
+  return datesAndPrices;
+};
+
+// NOTE: this date has timezone UTC, which is incorrect but works in this
+// case because we are just extracting the time
+const getAdjustedDateForTimestamp = (gmtoffset, timestamp) => {
+  const adjustedTimestamp =
+    (timestamp + gmtoffset) * constants.MILLISECONDS_PER_SECOND;
+  return new Date(adjustedTimestamp);
 };
 
-const padSingleDigitWithZero = value => {
-  let num = parseInt(value);
-  // need to check value because parseInt turns '12hello' into a number
-  if (isNaN(value) || isNaN(num)) {
-    throw new TypeError(`${padSingleDigitWithZero.name} requires a number or numeric string`);
+const getDateAndTime = (gmtoffset, timestamp, dateAndTimeFormat) => {
+  const dateAndTime = getAdjustedDateForTimestamp(gmtoffset, timestamp);
+  return dateFormat(dateAndTime, dateAndTimeFormat, true);
+};
+
+const getStartOfDayTimestampIndex = (dayIndex, timestampsPerDay) =>
+  Math.floor(dayIndex * timestampsPerDay);
+
+const getEndOfDayTimestampIndex = (dayIndex, timestampsPerDay) =>
+  Math.floor((dayIndex + 1) * timestampsPerDay);
+
+// Could be buggy depending on how the timestamps in their API are given.
+// Couldn't find info on this so will need to test at certain times of day.
+const getDatesAndTimesForOneDay = (
+  close,
+  dayIndex,
+  gmtoffset,
+  numberOfDays,
+  timestamp,
+  timestampIntervals
+) => {
+  let datesTimesAndPrices = [];
+  const timestampsPerDay = timestamp.length / numberOfDays;
+  const dateAndTimeFormat = (numberOfDays === constants.ONE_DAY)
+    ? constants.DATE_FORMAT_ONE_DAY
+    : constants.DATE_FORMAT_FIVE_DAY;
+
+  for (
+    let i = getStartOfDayTimestampIndex(dayIndex, timestampsPerDay);
+    i < getEndOfDayTimestampIndex(dayIndex, timestampsPerDay);
```

Comment:
having trouble wrapping my head around this right now, as I am a bit distracted.  Just want to confirm there's no off-by-one error here

Reply:
I'll think about it more tomorrow since you think there might be one, but when I thought it through when I wrote it there wasn't one